### PR TITLE
Custom esri token service

### PR DESCRIPTION
**Pull request by colin-geo**
# Overview
This PR updates the Esri provider so that a custom token generation service url can be passed through configuration to enable fetching of data from a self-hosted ArcGIS Server instance.

# Related Issue / discussion
Fixes geopython/pygeoapi#1812

# Additional information
None 

# Dependency policy (RFC2)

- [X] I have ensured that this PR meets [RFC2](https://pygeoapi.io/development/rfc/2) requirements

# Updates to public demo

- [X] I have ensured that breaking changes to the [pygeoapi master demo server](https://github.com/geopython/demo.pygeoapi.io) have been addressed
  - [X] https://github.com/geopython/demo.pygeoapi.io/blob/master/services/pygeoapi_master/local.config.yml

# Contributions and licensing

(as per https://github.com/geopython/pygeoapi/blob/master/CONTRIBUTING.md#contributions-and-licensing)

- [X] I'd like to contribute bugfix to pygeoapi. I confirm that my contributions to pygeoapi will be compatible with the pygeoapi license guidelines at the time of contribution
- [X] I have already previously agreed to the pygeoapi Contributions and Licensing Guidelines

### CODE REVIEW

File: pygeoapi/provider/esri.py [hunk 1]
```diff
@@ -62,8 +62,10 @@ def __init__(self, provider_def):
         self.crs = provider_def.get('crs', '4326')
         self.username = provider_def.get('username')
         self.password = provider_def.get('password')
+        self.token_url = provider_def.get('token_service')
+        self.token_referer = provider_def.get('referer')
```
Comment (webb-ben) [inline pygeoapi/provider/esri.py:66]:
```suggestion
        self.token_url = provider_def.get('token_service', ARCGIS_URL)
        self.token_referer = provider_def.get('referer', GENERATE_TOKEN_URL)
```

File: pygeoapi/provider/esri.py [hunk 2]
```diff
@@ -62,8 +62,10 @@ def __init__(self, provider_def):
...
         self.token = None
-
+
```
Comment (webb-ben) [inline pygeoapi/provider/esri.py:68]:
```suggestion

```

File: pygeoapi/provider/esri.py [hunk 3]
```diff
@@ -194,16 +196,16 @@ def login(self):
                 msg = 'Missing ESRI login information, not setting token'
                 LOGGER.debug(msg)
                 return
-
+
```
Comment (webb-ben) [inline pygeoapi/provider/esri.py:199]:
```suggestion

```

File: pygeoapi/provider/esri.py [hunk 4]
```diff
@@ -194,16 +196,16 @@ def login(self):
...
                 LOGGER.debug(msg)
                 return
-
+                
             params = {
                 'f': 'pjson',
                 'username': self.username,
                 'password': self.password,
-                'referer': ARCGIS_URL
+                'referer': (ARCGIS_URL if self.token_referer is None else self.token_referer)
```
Comment (webb-ben) [inline pygeoapi/provider/esri.py:204]:
```suggestion
                'referer': self.token_referer
```

File: pygeoapi/provider/esri.py [hunk 5]
```diff
@@ -194,16 +196,16 @@ def login(self):
...
             }
 
             LOGGER.debug('Logging in')
-            with self.session.post(GENERATE_TOKEN_URL, data=params) as r:
+            with self.session.post((GENERATE_TOKEN_URL if self.token_url is None else self.token_url), data=params) as r:
```
Comment (webb-ben) [inline pygeoapi/provider/esri.py:208]:
```suggestion
            with self.session.post(self.token_url, data=params) as r:
```

**Review by webb-ben — CHANGES_REQUESTED** [review]
Looks good. Small changes. Can we also add the new fields to the ESRI OAF Provider documentation (https://github.com/geopython/pygeoapi/blob/master/docs/source/data-publishing/ogcapi-features.rst#esri-feature-service)

**Comment by colin-geo** [comment]
I have incorporated the suggestions.

**Review by webb-ben — APPROVED** [review]

**Pull request merged**